lib/dialects/snowflake/query-generator.js

```diff
@@ -0,0 +1,633 @@
+      }
+    }
+
+    return template;
+  }
+
+  attributesToSQL(attributes, options) {
+    const result = {};
+
+    for (const key in attributes) {
+      const attribute = attributes[key];
+      result[attribute.field || key] = this.attributeToSQL(attribute, options);
+    }
+
+    return result;
+  }
+
+  /**
+   * Check whether the statmement is json function or simple path
+   *
+   * @param   {string}  stmt  The statement to validate
+   * @returns {boolean}       true if the given statement is json function
+   * @throws  {Error}         throw if the statement looks like json function but has invalid token
+   * @private
+   */
+  _checkValidJsonStatement(stmt) {
+    if (typeof stmt !== 'string') {
+      return false;
+    }
+
+    let currentIndex = 0;
+    let openingBrackets = 0;
+    let closingBrackets = 0;
+    let hasJsonFunction = false;
+    let hasInvalidToken = false;
+
+    while (currentIndex < stmt.length) {
+      const string = stmt.substr(currentIndex);
+      const functionMatches = JSON_FUNCTION_REGEX.exec(string);
+      if (functionMatches) {
+        currentIndex += functionMatches[0].indexOf('(');
+        hasJsonFunction = true;
+        continue;
+      }
+
+      const operatorMatches = JSON_OPERATOR_REGEX.exec(string);
+      if (operatorMatches) {
+        currentIndex += operatorMatches[0].length;
+        hasJsonFunction = true;
+        continue;
+      }
+
+      const tokenMatches = TOKEN_CAPTURE_REGEX.exec(string);
+      if (tokenMatches) {
+        const capturedToken = tokenMatches[1];
+        if (capturedToken === '(') {
+          openingBrackets++;
+        } else if (capturedToken === ')') {
+          closingBrackets++;
+        } else if (capturedToken === ';') {
+          hasInvalidToken = true;
+          break;
+        }
+        currentIndex += tokenMatches[0].length;
+        continue;
+      }
+
+      break;
+    }
+
+    // Check invalid json statement
+    if (hasJsonFunction && (hasInvalidToken || openingBrackets !== closingBrackets)) {
+      throw new Error(`Invalid json statement: ${stmt}`);
+    }
+
+    // return true if the statement has valid json function
+    return hasJsonFunction;
+  }
+
+  dataTypeMapping(tableName, attr, dataType) {
+    if (dataType.includes('PRIMARY KEY')) {
+      dataType = dataType.replace('PRIMARY KEY', '');
+    }
+
+    if (dataType.includes('SERIAL')) {
+      if (dataType.includes('BIGINT')) {
+        dataType = dataType.replace('SERIAL', 'BIGSERIAL');
+        dataType = dataType.replace('BIGINT', '');
+      } else if (dataType.includes('SMALLINT')) {
+        dataType = dataType.replace('SERIAL', 'SMALLSERIAL');
+        dataType = dataType.replace('SMALLINT', '');
+      } else {
+        dataType = dataType.replace('INTEGER', '');
+      }
+      dataType = dataType.replace('NOT NULL', '');
+    }
+
+    if (dataType.startsWith('ENUM(')) {
+      dataType = dataType.replace(/^ENUM\(.+\)/, this.pgEnumName(tableName, attr));
```

Comment:
I don't think trust this this regex - there's any other `)` after the closing one everything up to that will be replaced.